resolver/type_syntax.cc

```diff
@@ -585,6 +585,23 @@ TypeSyntax::ResultType TypeSyntax::getResultTypeAndBind(core::MutableContext ctx
 
             auto sym = maybeAliased.data(ctx)->dealias(ctx);
             if (sym.data(ctx)->isClassOrModule()) {
+                // this is a special-case because T_Array and co all have zero typeArguments() internally, but we want
+                // to subject them to the same checks we have down below if they get used with no type arguments at all
+                if (sym == core::Symbols::T_Hash() || sym == core::Symbols::T_Array() ||
+                    sym == core::Symbols::T_Set() || sym == core::Symbols::T_Range() ||
+                    sym == core::Symbols::T_Enumerable() || sym == core::Symbols::T_Enumerator()) {
+                    if (auto e = ctx.state.beginError(i->loc, core::errors::Resolver::GenericClassWithoutTypeArgs)) {
+                        e.setHeader("Malformed type declaration. Generic class without type arguments `{}`",
+                                    sym.show(ctx));
+                        if (sym == core::Symbols::T_Hash()) {
+                            // Hash is special because it has arity 3 but you're only supposed to write the first 2
+                            e.replaceWith("Add type arguments", i->loc, "{}[T.untyped, T.untyped]", i->loc.source(ctx));
+                        } else {
+                            e.replaceWith("Add type arguments", i->loc, "{}[T.untyped]", i->loc.source(ctx));
+                        }
+                    }
+                }
```

Comment:
There's a bit of replicated logic between this conditional and the one right below it, is there a way to merge the two, or would that just complicate matters?

Reply:
There's a way to merge the two, but it'd be a _little bit_ hairy. The differences are
1. the error messages above do not include a `T::` (because it's already present on the symbol) while the latter do (because it's not present on the symbol but needs to be there for the autocorrect), and
2. the synthetic type arguments above are made by knowing ahead-of-time what the arity of the generic class is, while the ones below are done with the size of `typeMembers()` (which is zero for the `T::Type` generics.)

I'll give it a try, and I can always roll back if it's too hairy.

Reply:
This doesn't seem too hairy to me, thanks for combining the cases!